gefserver/server/api.go

```diff
@@ -218,62 +238,78 @@ func (s *Server) buildImageHandler(w http.ResponseWriter, r *http.Request, e env
+			err = s.db.SetBuildState(buildID, db.NewBuildStateError("Failed while creating file to save file part "+part.FileName(), 1))
+			if err != nil {
+				log.Println(err)
+			}
 			return
 		}
 		defer dst.Close()
 
-		if _, err := io.Copy(dst, part); err != nil {
-			Response{w}.ServerError("while dumping file part ", err)
+		if _, err = io.Copy(dst, part); err != nil {
+			log.Print("while dumping file part ", err)
+			err = s.db.SetBuildState(buildID, db.NewBuildStateError("Failed while dumping file part "+part.FileName(), 1))
+			if err != nil {
+				log.Println(err)
+			}
 			return
 		}
 
 		if strings.HasSuffix(strings.ToLower(part.FileName()), ".tar") || strings.HasSuffix(strings.ToLower(part.FileName()), ".tar.gz") {
-			tarFileFound = true
-			foundImageFileName = part.FileName()
+			tarArchiveName = part.FileName()
 		}
 
 		if strings.ToLower(part.FileName()) == "dockerfile" {
-			dockerFileFound = true
+			hasDockerfile = true
 		}
 
 	}
 
-	// Building an image from a Dockerfile
-	if dockerFileFound {
-		if _, err := os.Stat(filepath.Join(buildDir, "Dockerfile")); os.IsNotExist(err) {
-			Response{w}.ServerError("no Dockerfile to build new image ", err)
-			return
+	// Building from tar
+	if tarArchiveName != "" && !hasDockerfile {
+		err = s.db.SetBuildState(buildID, db.NewBuildStateOk("Importing an image from a tar archive", -1))
+		if err != nil {
+			log.Println(err)
 		}
+		go s.pier.StartServiceBuildFromTar(buildID, buildDir, connectionID, user.ID, tarArchiveName)
+	}
 
-		service, err = s.pier.BuildService(connectionID, user.ID, buildDir)
+	// Building from a Dockerfile
+	if hasDockerfile {
+		err := s.db.SetBuildState(buildID, db.NewBuildStateOk("Building an image from a Dockerfile", -1))
 		if err != nil {
-			Response{w}.ServerError("build service failed: ", err)
-			return
+			log.Println(err)
 		}
-	} else {
-		// Importing an existing image from a tar archive
-		if tarFileFound {
-			log.Println("Docker image file has been detected, trying to import")
-			log.Println(filepath.Join(buildDir, foundImageFileName))
-			service, err = s.pier.ImportImage(connectionID, user.ID, filepath.Join(buildDir, foundImageFileName))
-			if err != nil {
-				Response{w}.ServerError("while importing a Docker image file ", err)
-				return
-			}
+		go s.pier.StartServiceBuildFromFile(buildID, buildDir, connectionID, user.ID)
+	}
 
-			log.Println("Docker image has been imported")
-		} else {
-			Response{w}.ServerNewError("there is neither Dockerfile nor Tar archive")
-			return
-		}
+	Response{w}.Ok(jmap("buildID", buildID))
+}
+
+func (s *Server) inspectBuildImageHandler(w http.ResponseWriter, r *http.Request, e environment) {
+	allow, user := Authorization{s, w, r}.allowInspectBuild()
+	if user == nil || !allow {
+		return
 	}
 
-	Response{w}.Ok(jmap("Service", service))
+	vars := mux.Vars(r)
+	buildID := vars["buildID"]
+
+	build, err := s.db.GetBuild(buildID)
+	if err != nil {
+		Response{w}.ClientError("cannot get a build", err)
```

Comment:
if the build is not found, it's a client error, but can also be a server error if the db request fails for some reason.